libraries/mbed/targets/hal/TARGET_STM/TARGET_STM32F4/serial_api.c

```diff
@@ -479,4 +631,537 @@ void serial_break_clear(serial_t *obj)
+#if DEVICE_SERIAL_ASYNCH_DMA
+    if ((UartHandle.Instance->CR3 & USART_CR3_DMAT) !=0) {
+        // call dma tx interrupt
+        HAL_DMA_IRQHandler(UartHandle.hdmatx);
+    }
+    if ((UartHandle.Instance->CR3 & USART_CR3_DMAR) !=0) {
+        // call dma rx interrupt
+        HAL_DMA_IRQHandler(UartHandle.hdmarx);
+    }
+#endif
+    HAL_UART_IRQHandler(&UartHandle);
+  // TX PART:
+    if (__HAL_UART_GET_FLAG(&UartHandle, UART_FLAG_TC) != RESET) {
+        __HAL_UART_CLEAR_FLAG(&UartHandle, UART_FLAG_TC);
+        // return event SERIAL_EVENT_TX_COMPLETE if requested
+        if ((SERIAL_OBJ(events) & SERIAL_EVENT_TX_COMPLETE ) != 0){
+            return_event |= SERIAL_EVENT_TX_COMPLETE & obj->serial.events;
+        }
+    }
+    // handle error events:
+    if (__HAL_UART_GET_FLAG(&UartHandle, HAL_UART_ERROR_PE)) {
+        __HAL_UART_CLEAR_FLAG(&UartHandle, HAL_UART_ERROR_PE);
+        return_event |= SERIAL_EVENT_RX_PARITY_ERROR & obj->serial.events;
+    }
+    if (__HAL_UART_GET_FLAG(&UartHandle, HAL_UART_ERROR_NE)) {
+      __HAL_UART_CLEAR_FLAG(&UartHandle, HAL_UART_ERROR_NE);
+      // not supported by mbed
+    }
+    if (__HAL_UART_GET_FLAG(&UartHandle, HAL_UART_ERROR_FE)) {
+      __HAL_UART_CLEAR_FLAG(&UartHandle, HAL_UART_ERROR_FE);
+        return_event |= SERIAL_EVENT_RX_FRAMING_ERROR & obj->serial.events;
+    }
+    if (__HAL_UART_GET_FLAG(&UartHandle, HAL_UART_ERROR_ORE)) {
+      __HAL_UART_CLEAR_FLAG(&UartHandle, HAL_UART_ERROR_ORE);
+        return_event |= SERIAL_EVENT_RX_OVERRUN_ERROR & obj->serial.events;
+    }
+
+    //RX PART
+    // increment rx_buff.pos
+    if (UartHandle.RxXferSize !=0) { 
+        obj->rx_buff.pos = UartHandle.RxXferSize - UartHandle.RxXferCount;
+    }
+    if ((UartHandle.RxXferCount==0)&&(obj->rx_buff.pos >= (obj->rx_buff.length - 1))) {
+        return_event |= SERIAL_EVENT_RX_COMPLETE & obj->serial.events;
+    }
+    if ((buf != NULL) && (buf[obj->rx_buff.pos-1] == obj->char_match) && (SERIAL_OBJ(events) & SERIAL_EVENT_RX_CHARACTER_MATCH)) {
+      return_event |= SERIAL_EVENT_RX_CHARACTER_MATCH & obj->serial.events;
+    }
+    return return_event;  
+}
+
+/** Abort the ongoing TX transaction. It disables the enabled interupt for TX and
+ *  flush TX hardware buffer if TX FIFO is used
+ *
+ * @param obj The serial object
+ */
+void serial_tx_abort_asynch(serial_t *obj)
+{
+    UartHandle.Instance = (USART_TypeDef *)(SERIAL_OBJ(uart));
+    __HAL_UART_DISABLE_IT(&UartHandle, UART_IT_TC|UART_IT_TXE);
+    UartHandle.Instance = 0;
+
+    obj->tx_buff.buffer = 0;
+    obj->tx_buff.length = 0;
+  
+}
+
+/** Abort the ongoing RX transaction It disables the enabled interrupt for RX and
+ *  flush RX hardware buffer if RX FIFO is used
+ *
+ * @param obj The serial object
+ */
+void serial_rx_abort_asynch(serial_t *obj)
+{
+    UartHandle.Instance = (USART_TypeDef *)(SERIAL_OBJ(uart));
+    __HAL_UART_DISABLE_IT(&UartHandle, UART_IT_RXNE);
+    UartHandle.Instance = 0;
```

Comment:
Is this needed ? Instance is always refreshed if any of functions here. Same applies to the below buffer & length?  What about DMA? _IT states it's API for interrupts?